kernel/kernel-api/src/main/java/io/delta/kernel/internal/TransactionBuilderImpl.java

```diff
@@ -163,6 +193,53 @@ public Transaction build(Engine engine) {
       }
     }
 
+    /* --------------- Handle Domain Metadata ------------ */
+    // Enable the feature if necessary
+    if (!TableFeatures.isDomainMetadataSupported(protocol)) {
+      if (!domainMetadatasAdded.isEmpty()) {
+        // This txn is setting a domain metadata, enable the feature in the protocol
+        logger.info(
+            "Automatically enabling writer feature: {}", DOMAIN_METADATA_W_FEATURE.featureName());
+        protocol =
+            protocol.withNewWriterFeatures(
+                Collections.singleton(DOMAIN_METADATA_W_FEATURE.featureName()));
+        shouldUpdateProtocol = true;
+      }
+      if (!domainMetadatasRemoved.isEmpty()) {
+        // Do nothing here, we will throw an exception when the domain does not exist in the
+        // current snapshot
+      }
+    }
+    // Generate the tombstones for the removed domain metadatas
+    Map<String, DomainMetadata> snapshotDomainMetadataMap = snapshot.getDomainMetadataMap();
+    for (String domainName : domainMetadatasRemoved) {
+      if (snapshotDomainMetadataMap.containsKey(domainName)) {
+        DomainMetadata domainToRemove = snapshotDomainMetadataMap.get(domainName);
+        // It is possible that domainToRemove is a tombstone in the snapshot (aka already removed).
+        // In this case we still add it to the metadata domains as removed so that it can be
+        // correctly resolved in the case of conflict resolution.
+        // This helps us to avoid failing in a scenario like the following...
+        // 1. V0 has domain "foo"
+        // 2. Connector reads V0, sees domain "foo", starts building txnA and calls
+        //    txnA.withDomainMetadataRemoved("foo")
+        // 3. Concurrent txn commits V1 and removes "foo" BEFORE connector calls txnA.build()
+        // 4. We would like to avoid throwing an error in this case! (when txnA.build() is called)
```

Comment:
so the issue is that txnA.build() will see a read snapshot of V1?

Reply:
yes, txnA has readSnapshot=V1, but connector has readSnapshot=V0